### ISSUE TITLE
Project 2 Level 4 Test case 4_4 :<

### ISSUE THREAD

**Issue by limyapkai** [nus-cs2030/2021-s1#465]
Hi! I seem to be stuck at this cruel piece of test case for 2 days... Tried many variations but to no avail. Code Crunch either return me this error... or worser results... Any insights to this matter ?

![image](https://user-images.githubusercontent.com/27536739/99878158-65a77580-2c3e-11eb-907c-acce305d8e26.png)

**Comment by azr124** [comment]
yes I am stuck at this level too! I got a time limit exceeded error but my result doesn't print an infinite loop. Would be great to have a sample input to debug.

**Comment by LapisRaider** [comment]
The input for test case4_4 is: 2 5 10 1000 7 1

I had a similar error as yours before, and if I remember correctly it was due to when I start generating the service time. If the order of the generation is different than that of the test case, it'll cause the error. For example, if customer 9 is supposed to generate it first, but for some reason, customer 8 generates the service time first instead.

Might be where you call the getServiceTime generation that's causing the problem? It should be called in ServeEvent. Hope this helps.

**Comment by crys-22** [comment]
Hmmm did you account for a clash in ServeEvent? i.e the availability from the server passed into it is not equal to the availability of the server with that same id when found using x.find() etc

**Comment by FergusMok** [comment]
For line 21 and 22, I don't think they can be served by the same server at the same time. Try to find a loophole in the logic there. Everything might solve itself once you solve this. 

Im thinking, it's because your serveEvents have the same timing in the priority queue. You should adjust the other serveEvents who have the same server, in the server's queue, when you have a execute this particular serveEvent.

**Comment by chongweil907** [comment]
[test4.4.txt](https://github.com/nus-cs2030/2021-s1/files/5577884/test4.4.txt)


For those who needs the sample output :)

**Comment by e0420658** [comment]
#301  hi you can try referring to this post i opened as i faced the same issue previously, but the gist of it its that you have customers being served at the same time meaning that you have to pull the updated server from the shop each time you assign it to wait/serve. additionally, you will have to check if the server next availtime coincides with the supposed "serve event", if the next available time is more, you would have to reschedule the serveEvent

**Comment by dljw98** [comment]
Hi, I faced this issue before also, for me it was an issue with the fact that the queue was not tallying and updating properly. Perhaps you'd want to System.out.println() the queue values to check if they are being updated correctly?

**Comment by limyapkai** [comment]
@FergusMok @LapisRaider @chongweil907 @e0420658  Thank you all for the comments! I had to reimplement the getServiceTime because I did not set it from the ServerEvent.... Anyway, I ran into ANOTHER deep problem sigh..

![image](https://user-images.githubusercontent.com/27536739/99901794-4a973d00-2cf4-11eb-801f-65e7319e21f3.png)

**Comment by FergusMok** [comment]
This problem should be solved quite easily. 
Check your order of the event comparator that you use in your priority queue. From the output, you can see that arrive > served > done. 
Also, the earlier customer's event should go first if there is a tie in the event type.
